File: windows/crates/router-core/src/engine/group_model.rs
//! Um grupo de contas que rotacionam entre si, e a configuração inteira do app
//! (`config.json`). Formato idêntico ao do macOS — é o contrato do PORTING.md.

use serde::{Deserialize, Serialize};

use super::account_model::Account;
use super::config_dir::ConfigDir;
use super::provider::Provider;
use crate::ids::Id;

pub const CURRENT_VERSION: i64 = 1;

fn default_threshold() -> f64 {
    90.0
}
fn default_true() -> bool {
    true
}
fn default_version() -> i64 {
    CURRENT_VERSION
}

/// Um grupo de contas que rotacionam entre si. Cada grupo tem o seu perfil e a
/// sua lista ordenada, então vários rotacionam em paralelo sem um pisar no outro.
#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AccountGroup {
    pub id: Id,
    pub name: String,
    #[serde(default)]
    pub provider: Provider,

    /// As contas na **ordem de preferência**: a primeira disponível é a escolhida.
    /// Chave `accountIDs` (com o "ID" maiúsculo, como o macOS grava).
    #[serde(rename = "accountIDs")]
    pub account_ids: Vec<Id>,

    /// O perfil onde as sessões deste grupo rodam. Um único grupo pode ser o
    /// padrão (sem variável de ambiente); os demais têm perfil dedicado.
    pub config_dir: ConfigDir,

    /// A partir de quanto de uso trocar de conta (por grupo).
    #[serde(default = "default_threshold")]
    pub threshold_percent: f64,

    /// Rotação automática ligada.
    #[serde(default = "default_true")]
    pub auto_rotate: bool,
}

impl AccountGroup {
    pub fn new(name: impl Into<String>, config_dir: ConfigDir) -> Self {
        AccountGroup {
            id: Id::new(),
            name: name.into(),
            provider: Provider::Anthropic,
            account_ids: Vec::new(),
            config_dir,
            threshold_percent: 90.0,
            auto_rotate: true,
        }
    }
}

/// A configuração inteira do app, persistida como um JSON só (`config.json`).
#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RouterConfig {
    #[serde(default)]
    pub accounts: Vec<Account>,
    #[serde(default)]
    pub groups: Vec<AccountGroup>,

    /// Um só histórico entre os grupos (`--resume` enxerga tudo), ou cada grupo
    /// com o seu. Padrão de fábrica é compartilhado.
    #[serde(default = "default_true")]
    pub share_history: bool,

    /// Versão do formato, para migrar sem perder configuração.
    #[serde(default = "default_version")]
    pub version: i64,
}

impl Default for RouterConfig {
    fn default() -> Self {
        RouterConfig {
            accounts: Vec::new(),
            groups: Vec::new(),
            share_history: true,
            version: CURRENT_VERSION,
        }
    }
}

impl RouterConfig {
    pub fn account(&self, id: Id) -> Option<&Account> {
        self.accounts.iter().find(|a| a.id == id)
    }

    /// As contas de um grupo, na ordem de preferência, ignorando IDs órfãos.
    pub fn accounts_in(&self, group: &AccountGroup) -> Vec<&Account> {
        group
            .account_ids
            .iter()
            .filter_map(|id| self.accounts.iter().find(|a| a.id == *id))
            .collect()
    }

    /// O grupo marcado como padrão — o que roda no `~/.claude`. No máximo um.
    pub fn default_group(&self) -> Option<&AccountGroup> {
        self.groups.iter().find(|g| g.config_dir.is_default)
    }
}

File: macos/Sources/CCUsageCore/Engine/GroupModel.swift
import Foundation

/// Um grupo de contas que rotacionam entre si.
///
/// É o coração do que o produto faz e o `cswap` não faz: cada grupo tem o seu
/// perfil e a sua lista ordenada, então vários grupos rotacionam em paralelo —
/// trabalho, pessoal, faculdade — sem um pisar no outro. O usuário cria, nomeia,
/// ordena e liga/desliga cada um pela UI.
public struct AccountGroup: Sendable, Equatable, Codable, Identifiable {
    public let id: UUID
    public var name: String
    public let provider: Provider

    /// As contas na **ordem de preferência** que o usuário definiu: a primeira
    /// disponível é a escolhida. Arrastar na UI reescreve esta ordem.
    public var accountIDs: [UUID]

    /// O perfil onde as sessões deste grupo rodam. Um único grupo pode ser o
    /// padrão (`~/.claude`, sem variável de ambiente); os demais têm perfil
    /// dedicado. Configurável — qual grupo é o padrão é escolha do usuário.
    public var configDir: ConfigDir

    /// A partir de quanto de uso trocar de conta. Por grupo, porque um grupo
    /// compartilhado com colegas pode querer folga maior que um pessoal.
    public var thresholdPercent: Double

    /// Rotação automática ligada. Desligado, o grupo ainda existe e troca na
    /// mão, mas não sozinho.
    public var autoRotate: Bool

    public init(id: UUID = UUID(), name: String, provider: Provider = .anthropic,
                accountIDs: [UUID] = [], configDir: ConfigDir,
                thresholdPercent: Double = 90, autoRotate: Bool = true) {
        self.id = id
        self.name = name
        self.provider = provider
        self.accountIDs = accountIDs
        self.configDir = configDir
        self.thresholdPercent = thresholdPercent
        self.autoRotate = autoRotate
    }
}

/// A configuração inteira do app, persistida como um JSON só.
///
/// Tudo que o usuário monta pela UI vive aqui: as contas, os grupos, a ordem, os
/// limiares, e as preferências que na primeira versão foram perguntas de
/// setup — qual grupo é o padrão, se o histórico é compartilhado. Nenhuma delas
/// é constante de código; o que o app traz de fábrica é só o valor inicial.
public struct RouterConfig: Sendable, Equatable, Codable {
    public var accounts: [Account]
    public var groups: [AccountGroup]

    /// Um só histórico de conversas entre os grupos (`--resume` enxerga tudo),
    /// ou cada grupo com o seu. Configurável; padrão de fábrica é compartilhado.
    public var shareHistory: Bool

    /// Versão do formato, para migrar sem perder configuração de quem já usava.
    public var version: Int

    public static let currentVersion = 1

    public init(accounts: [Account] = [], groups: [AccountGroup] = [],
                shareHistory: Bool = true, version: Int = RouterConfig.currentVersion) {
        self.accounts = accounts
        self.groups = groups
        self.shareHistory = shareHistory
        self.version = version
    }

    public func account(_ id: UUID) -> Account? { accounts.first { $0.id == id } }

    /// As contas de um grupo, na ordem de preferência, ignorando IDs órfãos.
    public func accounts(in group: AccountGroup) -> [Account] {
        group.accountIDs.compactMap { id in accounts.first { $0.id == id } }
    }

    /// O grupo marcado como padrão — o que roda em `~/.claude`. No máximo um.
    public var defaultGroup: AccountGroup? { groups.first { $0.configDir.isDefault } }
}
